pkg-r/vignettes/data-sources.Rmd

```diff
@@ -0,0 +1,148 @@
+vignette: >
+  %\VignetteIndexEntry{Data Sources}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>",
+  eval = FALSE
+)
+```
+
+`querychat` supports different types of data sources, including:
+
+1. Data frames
+2. DBI database connections (e.g., SQLite, PostgreSQL, MySQL, DuckDB)
+3. Custom `DataSource` interfaces
+
+The sections below describe how to use each type of data source with `querychat`.
+
+## Data frames
+
+You can use any data frame as a data source in `querychat`. Simply pass it to `QueryChat$new()`:
+
+```{r}
+library(querychat)
+library(palmerpenguins)
+
+querychat_app(mtcars)
+```
+
+Behind the scenes, `querychat` creates an in-memory DuckDB database and registers your data frame as a table for SQL query execution.
+
+## Database connections
+
+You can also connect `querychat` directly to any database supported by [DBI](https://dbi.r-dbi.org/). This includes popular databases like SQLite, DuckDB, PostgreSQL, MySQL, and many more.
+
+Assuming you have a database set up and accessible, you can create a DBI connection and pass it to `QueryChat$new()`. Below are some examples for common databases.
+
+### DuckDB
+
+```{r}
+library(DBI)
+library(duckdb)
+library(querychat)
+
+# Connect to a DuckDB database file
+con <- dbConnect(duckdb::duckdb(), dbdir = "my_database.duckdb")
+
+querychat_app(con, "my_table")
+
+# Don't forget to disconnect when done
+# dbDisconnect(con)
+```
+
+### SQLite
+
+```{r}
+library(DBI)
+library(RSQLite)
+library(querychat)
+
+# Connect to a SQLite database file
+con <- dbConnect(RSQLite::SQLite(), "my_database.db")
+
+querychat_app(con, "my_table")
+
+# Don't forget to disconnect when done
+# dbDisconnect(con)
+```
+
+### PostgreSQL
+
+```{r}
+library(DBI)
+library(RPostgres)
+library(querychat)
+
+# Connect to PostgreSQL
+con <- dbConnect(
+  RPostgres::Postgres(),
+  host = "localhost",
+  port = 5432,
+  dbname = "mydatabase",
+  user = "myuser",
+  password = "mypassword"
+)
+
+querychat_app(con, "my_table")
+
+# Don't forget to disconnect when done
+# dbDisconnect(con)
+```
+
+### MySQL
+
+```{r}
+library(DBI)
+library(RMariaDB)
+library(querychat)
+
+# Connect to MySQL
+con <- dbConnect(
+  RMariaDB::MariaDB(),
+  host = "localhost",
+  port = 3306,
+  dbname = "mydatabase",
+  user = "myuser",
+  password = "mypassword"
+)
+
+querychat_app(con, "my_table")
+
+# Don't forget to disconnect when done
+# dbDisconnect(con)
+```
+
+## Creating a database from a data frame
+
+If you don't have a database set up, you can easily create a local DuckDB database from a data frame:
+
+```{r}
+library(DBI)
+library(duckdb)
+
+con <- dbConnect(duckdb::duckdb(), dbdir = "my_database.duckdb")
+
+# Write a data frame to the database
+dbWriteTable(con, "penguins", penguins)
+
+# Or from CSV
+dbExecute(con, "
+  CREATE TABLE my_table AS
+  SELECT * FROM read_csv_auto('path/to/your/file.csv')
+")
```

Comment:
I'm pretty certain the duckdb R package has a friendlier function that directly does this